sdks/python/apache_beam/io/datastore/v1/helper.py

```diff
@@ -82,3 +94,148 @@ def str_compare(s1, s2):
+  """Returns a Cloud Datastore client."""
+
+  credentials = googledatastore.helper.get_credentials_from_env()
+  datastore = Datastore(project, credentials)
+  return datastore
+
+
+def make_request(project, namespace, query):
+  """Make a Cloud Datastore request for the given query."""
+
+  req = datastore_pb2.RunQueryRequest()
+  req.partition_id.CopyFrom(make_partition(project, namespace))
+
+  req.query.CopyFrom(query)
+  return req
+
+
+def make_partition(project, namespace):
+  """Make a PartitionId for the given project and namespace."""
+
+  partition = entity_pb2.PartitionId()
+  partition.project_id = project
+  if namespace is not None:
+    partition.namespace_id = namespace
+
+  return partition
+
+
+def retry_on_rpc_error(exception):
+  """A retry filter for Cloud Datastore RPCErrors."""
+
+  if isinstance(exception, RPCError):
+    if exception.code >= 500:
+      return True
+    else:
+      return False
+  else:
+    # TODO(vikasrk): Figure out what other errors should be retried.
+    return False
+
+
+def fetch_entities(project, namespace, query, datastore):
+  """A helper method to fetch entities from Cloud Datastore.
+
+  Args:
+    project: Project ID
+    namespace: Cloud Datastore namespace
+    query: Query to be read from
+    datastore: Cloud Datastore Client
+
+  Returns:
+    An iterator of entities.
+  """
+  return QueryIterator(project, namespace, query, datastore)
+
+
+def make_latest_timestamp_query(namespace):
+  """Make a Query to fetch the latest timestamp statistics."""
+
+  query = query_pb2.Query()
+  if namespace is None:
+    query.kind.add().name = '__Stat_Total__'
+  else:
+    query.kind.add().name = '__Stat_Ns_Total__'
+
+  # Descending order of `timestamp`
+  datastore_helper.add_property_orders(query, "-timestamp")
+  # Only get the latest entity
+  query.limit.value = 1
+  return query
+
+
+def make_kind_stats_query(namespace, kind, latest_timestamp):
+  """Make a Query to fetch the latest kind statistics."""
+
+  kind_stat_query = query_pb2.Query()
+  if namespace is None:
+    kind_stat_query.kind.add().name = '__Stat_Kind__'
+  else:
+    kind_stat_query.kind.add().name = '__Stat_Ns_Kind__'
+
+  kind_filter = datastore_helper.set_property_filter(
+      query_pb2.Filter(), 'kind_name', PropertyFilter.EQUAL, unicode(kind))
+  timestamp_filter = datastore_helper.set_property_filter(
+      query_pb2.Filter(), 'timestamp', PropertyFilter.EQUAL,
+      latest_timestamp)
+
+  datastore_helper.set_composite_filter(kind_stat_query.filter,
+                                        CompositeFilter.AND, kind_filter,
+                                        timestamp_filter)
+  return kind_stat_query
+
+
+class QueryIterator(object):
+  """A iterator class for entities of a given query.
+
+  Entities are read in batches. Retries on failures.
+  """
+
+  _NOT_FINISHED = query_pb2.QueryResultBatch.NOT_FINISHED
+  # Maximum number of results to request per query.
+  _BATCH_SIZE = 500
```

Comment:
I thin this is a limit enforced by Datastore, right ? If so please mention that.